### ISSUE TITLE
A/P sustained load test

### ISSUE THREAD

**Issue by ahus1** [keycloak/keycloak-benchmark#459]
A/P, one Region, 2 independent OpenShift clusters in two different AZs, 1 Aurora DB (don’t care about AZ), 100_000 users, 300 logins per second, 100_000 active sessions 

Sustained load run auf authorization code, see if sessions are replicated, check errors in log

Ideas (might be out of scope): 
* Maybe reduce the number of owners from 2 to 1 as entries are safe in the remote Infinispan. 
Possible impact: Faster write (fewer updates to send to other nodes), slower read (as less sessions are local)
* Maybe reduce the number of entries in a cache (eviction) as entries are safe in the remote Infinispan
Possible impact: Less memory use in Keycloak even for large installations, slower reads for those entries not in the cache. 

For reads: Primary owner will read from the remote store if it doesn't have it in memory. Backup owner will go to the remote store if it hasn't it in memory. Non-Owner would go to primary owner. 

For a replicated store, everyone is the owner of the keys; so an idea: Make all caches replicated with an eviction policy. If they don't find the entry in memory for reads, they will directly read from the remote store. On write, they will need to update all Keycloak instances. Maybe this is better for 3 Keycloak pods, but it might be less efficient for 6 or 9 Keycloak pods.

**Comment by ahus1** [comment]
Currently blocked by the investigate of data corruption in cross-DC communication

**Comment by kami619** [comment]
We(myself and Michal) are looking at the patch image provided by @pruivo to see if we are able to get across the data corruption issue.

**Comment by kami619** [comment]
using the patch image we were able to overcome the initial problems with load on a cross data center setup. 

However, we ran into pod restarts with liveness and readiness probes restarting the Keycloak pods during the sustained load test run with 200 users-per-sec. Then we tried to disable the Keycloak pods' liveness probe and readiness probe and repeated the sustained load test, and below are some high level outcomes and the cluster configuration used. 

Keycloak Cluster config:

```
#ISPN
KC_CUSTOM_INFINISPAN_CONFIG_FILE=config/kcb-infinispan-cache-remote-store-config.xml
KC_ISPN_NAMESPACE=mhajas-keycloak


#DB
KC_DATABASE=aurora-postgres
KC_DATABASE_URL=mhajas.cluster-clhthfqe0h8p.eu-west-1.rds.amazonaws.com
KC_DB_POOL_MIN_SIZE=15
KC_DB_POOL_MAX_SIZE=15
KC_DB_POOL_INITIAL_SIZE=15
KC_STORAGE=

#KC
KC_INSTANCES=6
KC_DISABLE_STICKY_SESSION='true'
KC_NAMESPACE_PREFIX='mhajas-'
KC_MEMORY_REQUESTS_MB=2048
KC_MEMORY_LIMITS_MB=3000
KC_HEAP_MAX_MB=2048
KC_HEAP_INIT_MB=1024
KC_CPU_REQUESTS=8
KC_CPU_LIMITS=8

```


with 200 users-per-sec for the Authorization Code scenario, liveness and readiness probes disabled.

![Image](https://github.com/keycloak/keycloak-benchmark/assets/3415105/88601728-3c28-4538-88ae-fb91b574692a)

- We observed we had good throughput of 185 reqs/sec per endpoint
- Mean and 50th percentile response times are comparable
- No interruption to the availability of services
- Spike in response times, mostly related to the increased latency in replication time between ISPN cross-site


![Image](https://github.com/keycloak/keycloak-benchmark/assets/3415105/0843b03e-07c3-4410-970a-136d0dbdf860)


with 300 users-per-sec for the Authorization Code scenario, liveness and readiness probes disabled.



![Image](https://github.com/keycloak/keycloak-benchmark/assets/3415105/57ceeb48-693e-4a5a-b9fc-011ae9819942)


- We didn't see comparable response times or throughput in relation to a single OCP cluster based sustained load run
- Developers ascertained an issue with how internal cross-site is currently working and some thoughts are being collated here - https://github.com/keycloak/keycloak-benchmark/issues/473

**Comment by ahus1** [comment]
* Test with single gossip router and ensure that results with 200 users per second are reproducible
* Test with 300 users per second
* If test with 300 users per second fails, find bottleneck

**Comment by mhajas** [comment]
I can confirm the run with 200 users per second is reproducible also with a single gossip router.

Results:
![image](https://github.com/keycloak/keycloak-benchmark/assets/13901995/08254153-b8e3-44ad-ad6b-00ef3e054377)

There were no peaks during the run and replication time to the second DC stayed basically consistent across the whole test:
![image](https://github.com/keycloak/keycloak-benchmark/assets/13901995/7951550e-98d1-42b0-9698-67e2d37ad8bd)

**Comment by mhajas** [comment]
Running sustained load for 300UPS passed as well. I was not able to reproduce failures encountered by @kami619 

Here are the results:
![image](https://github.com/keycloak/keycloak-benchmark/assets/13901995/285cc2f0-b986-460f-b788-244df9bf6802)


I tried to increase UPS to 350 and it passed too:
![image](https://github.com/keycloak/keycloak-benchmark/assets/13901995/2f88d806-e629-4456-ba54-37b8efecc89f)

**Comment by pruivo** [comment]
One more, did we test the maximum users/sec that the active-passive can handle?

**Comment by mhajas** [comment]
> @mhajas I have a couple of questions
> 
> * Are these tests run with SYNC or ASYNC cross-site?

With SYNC, we wanted to find out how SYNC fail before we move to ASYNC.

> * Are the clusters in different regions/AZ?

We used different AZ within the same region

> One more, did we test the maximum users/sec that the active-passive can handle?

No, we reached 350 UPS which was enough for us at this stage.
